### ISSUE TITLE
Plans for Windows support?

### ISSUE THREAD

**Issue by cosmicdan** [beehive-lab/TornadoVM#16]
Hi,

Loving the project - something I've been looking for a long time to learn GPGPU programming. I'm very surprised it took me so long to hear about the work you guys have done.

I have a (long-term) goal to use TornadoVM for development of a videogame and simulations (I'm only a hobbyist for now, nothing commercial) so naturally I'm curious if there's any plans to bring Tornado to Windows since most of my intended audience would be using it.

Are there any known roadblocks for Windows or is it just something that hasn't been looked at yet?

Just to clarify - I only use Windows myself as a secondary OS so this is by no means urgent - just curious on if it's an option for future.

Thanks!

**Comment by jjfumero** [comment]
Hi, 
  This is something we haven't looked yet.  We do support Linux and OSx. However, since TornadoVM 0.6, we can also run with Graal 19.3 and JDK 11, and I think that combination could allow us to run on Windows (with a bit of tune from the bash scripts), but we haven't checked. This is in our future plans (long term).

**Comment by cosmicdan** [comment]
Thanks Juan, I'm quite satisfied with that as I wont targeting the Windows platform for a short while either.

If I do end up getting bored one day and experiment with Windows before you guys get to it, I'll report my findings here - since you made it sound quite trivial there. I was mostly worried about natives not being ported to Windows - not the case here? I admit I've not examined the TVM stack in detail, apologies.

Finally, GraalVM 20 has "proper" Windows support though, it's still considered early access in 19.3 from what I remember - so for sanity's sake I'd guess that getting TVM working under Graal 20 should be a prerequisite for Windows testing.

⁣Get BlueMail for Android ​

**Comment by jjfumero** [comment]
Thank you, 
  there are few parts with some dependencies like `cmake` to compile the JNI part of TornadoVM. But I think there is a binary for Windows. 

We do plan to add support for Graal 20.0. Stay tuned!

**Comment by vsilaev** [comment]
I was able to compile graal-jd8 and graal-jdk11 on Windows 10 Pro x64 with MingW64 and minor changes. I guess, any Win7+ x64 will work.

I've started with [v0.8](https://github.com/beehive-lab/TornadoVM/tree/v0.8), and noticed that code was upgraded to C++ version of JNI libs recently. So the patches should be cherry-picked later. Obviously, if project maintainers will find them useful. 

Attached are 2 Git patch files but what is more important is a guideline how to apply them (i.e. configure MSys2 + MingW64 correctly). Will add a textual description here in 1-3 days.

Disclaimer: only OpenCL build is executed, can't understand how PTX profile is activated, though I had specified both backends on build command line.

[TornadoVM-0.8-win64-build-patches.zip](https://github.com/beehive-lab/TornadoVM/files/5811663/TornadoVM-0.8-win64-build-patches.zip)
![proof-screen](https://user-images.githubusercontent.com/3337603/104527995-2d595d80-5617-11eb-9a61-37343358703d.png)

**Comment by jjfumero** [comment]
Thank you @vsilaev . This is great. Did you try the unittests? 

```bash
tornado-test.py -V 
```

Additionally, I suggest opening a PR, so it is easy for us to track. If possible, it would be good to add documentation of all dependencies and install process here:
https://github.com/beehive-lab/TornadoVM/tree/master/assembly/src/docs 


Regarding the PTX backend profile, @gigiblender can help here.

**Comment by vsilaev** [comment]
Juan,

I've tried a lot of examples - all works as expected. I have just re-run test suit: 327 tests passed, 4 ones failed (all failed tests are related to VirtualDevice).

![tests](https://user-images.githubusercontent.com/3337603/104564737-551ee480-565c-11eb-9d61-4f2ffb9da1d5.png)

I will add textual description here first, then I will create separate PR.

Btw, your C code is absolutely portable besides one single place where you used *NIX-specific `posix_memalign` function.

**Comment by jjfumero** [comment]
I think those unittests are expected to fail. We check with a pre-defined output. 

When we run the unittests, we create a log file `tornado_unittests.log`. You can send this log to us. 

Good to know the code is portable. Hope we can get the same with the C++ port.

**Comment by vsilaev** [comment]
[tornado_unittests.log](https://github.com/beehive-lab/TornadoVM/files/5813296/tornado_unittests.log)
Here is the full test log.

**Comment by gigiblender** [comment]
> 
> Disclaimer: only OpenCL build is executed, can't understand how PTX profile is activated, though I had specified both backends on build command line.
> 

@vsilaev the way we currently build Tornado with the PTX backend enabled is by running `make BACKEND=ptx`. You can also have both backends enabled (eg: `make BACKEND=opencl,ptx`). 
If you use a different JDK distribution than the default, then the build command line becomes, for example, `make graal-jdk-11 BACKEND=ptx,opencl`

If you have the time, do you mind sharing with us the output of building Tornado with the PTX backend enabled and the output of `tornado --devices` ? 

A reason for not being able to build/run the PTX backend might be because the Nvidia CUDA drivers are not setup properly.

**Comment by vsilaev** [comment]
@gigiblender 
I have [NVIDIA CUDA Toolkit](https://developer.nvidia.com/cuda-downloads?target_os=Windows&target_arch=x86_64&target_version=10) installed. 
But CUDA doesn't add any folders to default (USER/SYSTEM) `PATH` on Windows. Instead, it creates separate ENV variables - probably they are supposed to be added to the `PATH`. 

I do run `make graal-jdk-8 BACKEND={ptx,opencl}` but only `opencl` build is triggered and I haven't figured out (yet) how `ptx` is triggered...

**Comment by vsilaev** [comment]
I even make `ptx-backend` profile in corresponding `pom.xml` (from drivers) as active by default - still nothing...

**Comment by gigiblender** [comment]
> Wow! The order of params matters somehow:
> ` make graal-jdk-8 BACKEND={opencl,ptx}` -- PTX is included in build
> ` make graal-jdk-8 BACKEND={ptx,opencl}` -- PTX is not included in build

@vsilaev, you should not use the curly brackets in the build command. `make graal-jdk-8 BACKEND=opencl,ptx` should work. If it doesn't, then this is something that breaks on windows only and must be looked into.

**Comment by vsilaev** [comment]
Well, actually brackets are mentioned in the current documentation - [Install with Graal VM](https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/10_INSTALL_WITH_GRAALVM.md).

Removing them indeed helps! 
I build successfully PTX-only backend and re-run tests. Here are the results: 
[tornado_unittests (ptx).log](https://github.com/beehive-lab/TornadoVM/files/5814469/tornado_unittests.ptx.log)
(teaser: all tests passed)

**Comment by jjfumero** [comment]
> [tornado_unittests.log](https://github.com/beehive-lab/TornadoVM/files/5813296/tornado_unittests.log)
> Here is the full test log.

It seems the error is related to the format of the file. Probably because of the end of the line in Windows. We will fix this.

**Comment by vsilaev** [comment]
The most recent patch to enable build on Windows x64:
[0001-Enable-MingW64-build-on-windows-fi-Unixx-only-C-func.zip](https://github.com/beehive-lab/TornadoVM/files/5820148/0001-Enable-MingW64-build-on-windows-fi-Unixx-only-C-func.zip)

Instructions will follow...

**Comment by vsilaev** [comment]
# Build TornadoVM with Graal 20.2.0 JDK 8 / 11 on Windows x64
## Install tools
### Maven
Download Apache Maven from the [official site](https://maven.apache.org/download.cgi) and extract it to the any location. Below it's assumed that Maven's home is `C:/Maven`.
### MSys2
1. Download [MSys2](https://www.msys2.org/) installer from the official site and run it. You may choose any installation directory, below it will be referred as `<MSYS2>`. IMPORTANT: the only executable you should use as a terminal is `<MSYS2>/mingw64.exe`, don't use others (msys2.exe, mingw32.exe) as this would lead to unpredictable results.

2. Update MSys2 _**system**_ packages. Start `<MSYS2>/mingw64.exe` and run the following command in the terminal:
```bash
pacman -Syu
```
You might need to execute it several times until you see that no updates found.

3. Update MSys2 _**default**_ packages. In the terminal window of `<MSYS2>/mingw64.exe` run:
```bash
pacman -Su
```
You might need to execute it several times until you see that no updates found.

4. Install necessary tools to MSys2. In the terminal window of `<MSYS2>/mingw64.exe` run:
```bash
pacman -S \
mingw-w64-x86_64-make           \
mingw-w64-x86_64-cmake          \
mingw-w64-x86_64-gcc            \
mingw-w64-x86_64-opencl-headers \
mingw-w64-x86_64-opencl-icd-git \
python make git
```

5. Create default Maven repository for MSys2 user:
```bash
cd ~
mkdir .m2
```

6. Create file `mvn` in `<MSYS2>/mingw64/bin` with any textual editor according to the template below:
```bash
#!/usr/bin/env bash
BASH_FILE=$(where bash)
SETTINGS=\"${BASH_FILE%\\*\\*\\*}${HOME}/.m2/settings.xml\"

CMAKE_FILE=$(where cmake | head -n 1)
export CMAKE_ROOT=\"${CMAKE_FILE%\\*}\"
echo $CMAKE_ROOT

C:/Maven/bin/mvn.cmd --settings ${SETTINGS} "$@"
```
You should only change path to Maven executable to the actual location of Maven in your system.

## Get TornadoVM sources
Clone TornadoVM 0.8 from Git repository using `<MSYS2>/mingw64.exe`:
```
cd D:/MyProjects
git clone -b master --single-branch https://github.com/beehive-lab/TornadoVM.git
cd TornadoVM
git checkout tags/v0.8
```
We will refer hereafter the directory with TornadoVM sources as `<TornadoVM>`.

Now download patch to fix TornadoVM build on Windows and apply it.
[0001-Enable-MingW64-build-on-windows-fi-Unixx-only-C-func.zip](https://github.com/beehive-lab/TornadoVM/files/5820148/0001-Enable-MingW64-build-on-windows-fi-Unixx-only-C-func.zip)

## Get Graal 20.2.0
Visist [GraalVM Community Edition 20.2.0](https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-20.2.0) and download either of the following builds:
- [graalvm-ce-java8-windows-amd64-20.2.0.zip](https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-20.2.0/graalvm-ce-java8-windows-amd64-20.2.0.zip) -- Graal JDK 8 x64
- [graalvm-ce-java11-windows-amd64-20.2.0.zip](https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-20.2.0/graalvm-ce-java11-windows-amd64-20.2.0.zip) -- Graal JDK 11 x64

Extract the downloaded file to any directory.

## Install NVIDIA CUDA and NVIDIA driver
If you don't need PTX backend that depends on CUDA you may just download and install regular driver that corresponds to your GPU from [NVIDIA Driver Downloads](https://www.nvidia.com/Download/index.aspx) page. 

Otherwise you need complete CUDA Toolkit from [CUDA Toolkit Downloads](https://developer.nvidia.com/cuda-downloads?target_os=Windows&target_arch=x86_64) page. Note, that toolkit includes GPU drivers as well, so you may avoid downloading drivers separately. The only thing to note is that GPU driver you are currently using should be of the same or higher version than the one shipped with CUDA Toolkit. So if you have existing driver make sure that it's version is same or higher, otherwise update GPU driver during toolkit installation. Note, that NSight, BLAST libs and Visual Studio integration are irrelevant for TornadoVM builds, you need just SDK - so you may skip installing them.

## Configuring build
Using any text editor create file `<TornadoVM>/etc/sources.env`  with the following content:
```bash
#!/bin/bash

# UPDATE PATH TO ACTUAL LOCATION OF THE JDK OR GRAAL (REMEMBER OF UNIX_STYLE SLASHES AND SPACES!!!)
export JAVA_HOME="C:/graalvm-ce-java8-20.2.0"  

## NEXT TWO LINES NECESSARY TO BUILD PTX (NVIDIA CUDA) BACKEND
## COMMENT THEM OUT OR JUST IGNORE IF YOU ARE NOT INTERESTED IN PTX BUILD
## OTHERWISE UPDATE 'CUDA_PATH' WITH ACTUAL VALUE (REMEMBER OF UNIX_STYLE SLASHES AND SPACES!!!)
export CUDA_PATH="C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.2"
export PTX_LDFLAGS=-L\"$CUDA_PATH/lib/x64\"

# LEFT THE REST OF FILE 'AS IS'
# DON'T ALTER!
export PATH=$PWD/bin/bin:$PATH               ## This directory will be automatically generated during Tornado compilation
export TORNADO_SDK=$PWD/bin/sdk              ## This directory will be automatically generated during Tornado compilation
CMAKE_FILE=$(where cmake | head -n 1)
export CMAKE_ROOT=${CMAKE_FILE%\\*\\*}
```
There are only 2 places you should adjust:
1. `JAVA_HOME` path that points to your Graal installation
2. `CUDA_PATH` pointing to your NVIDIA GPU Computing Toolkit (CUDA) - this one is necessary only for builds with PTX backend.

## Running build
Start `<MSYS2>/mingw64.exe` terminal, navigate to the `<TornadoVM>`  directory and execute build:
```bash
cd D:/MyProjects/TornadoVM
source etc/sources.env
make graal-jdk-8 BACKEND=ptx,opencl
```
The latest command, `make`, depends on the Graal version you are using:
1. For Graal JDK 8 use `make graal-jdk-8 BACKEND=ptx,opencl` (as above)
2. For Graal JDK 11 use `make graal-jdk-11 BACKEND=ptx,opencl` 

The `BACKEND` parameter has to be a comma-separated list of `ptx` and `opencl` options. You may build `ptx` only when NVIDIA GPU Computing Toolkit (CUDA) is installed.

## Verifying build
Don't close `<MSYS2>/mingw64.exe` after build. Run the following command to see that TornadoVM is working:
```bash
tornado --devices
```
You should see a list of OpenCL and/or CUDA devices available on your system.

Now try to run a simple test. For Graal JDK 11:
```bash 
tornado -m tornado.examples/uk.ac.manchester.tornado.examples.compute.MatrixMultiplication2D 512
```

... for Graal JDK 8:
```bash 
tornado uk.ac.manchester.tornado.examples.compute.MatrixMultiplication2D 512
```

If "sanity checks" from above work, you can execute complete test suite for TornadoVM:
```bash
tornado-test.py -V
```

## Known issues
1. If you already have MSys2 installed and heavily customized you may experience issues with build or tests. We are suggesting to start with fresh MSys2 installation in this case and follow the instructions above. Most notably, make sure that you have no `mingw-w64-x86_64-python` installed - it prevents Python scripts that execute tests from running. Also, make sure that you have updated all GCC / Make / CMake packages mentioned.
2. If you see no output from `tornado --devices` this may be either of 2 reasons:
-- OpenCL / CUDA is misconfigured. Download any third-party tool for OpenCL / CUDA capabilities viewing and check that you can see your devices there. Sometimes order of installation of different OpenCL drivers matters - Intel OpenCL SDK may shadow NVIDIA OpenCL and alike.
-- You build native code of the library using wrong compiler, most probably you ran `<MSYS2>/msys2.exe` terminal instead of `<MSYS2>/mingw64.exe` . Please re-try with correct terminal (and therefore GCC) version.
3. If you see JVM crashes or `UnsatisfiedLinkError` or some `Error initializing DLL` during `tornado --devices` execution than it's definitely due to wrong GCC (and hence terminal) version used during build.

**Comment by vsilaev** [comment]
Hello,

I'm trying to adjust current `develop` branch code to support builds on Windows x64 with Mingw64/GCC
OpenCL JNI conversion ran smoothly, but I've staucked with the `ptx-jni` -- there is an error (+warnings) in `PTStream.cpp`

```
D:/Projects/TornadoVM/drivers/ptx-jni/target/windows-amd64-release/sources/source/PTXStream.cpp:222:40: warning: invalid conversion from 'const int*' to 'const jint*' {aka 'const long int*'} [-fpermissive]
  222 |             length / sizeof(JAVATYPE), static_cast<const JAVATYPE *>(staging_list->staging_area)); \
      |                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                        |
      |                                        const int*
```

 Then

```
D:/Projects/TornadoVM/drivers/ptx-jni/target/windows-amd64-release/sources/source/PTXStream.cpp:403:101: warning: invalid conversion from 'int*' to 'jint*' {aka 'long int*'} [-fpermissive]
```

And finally the error:

```
D:/Projects/TornadoVM/drivers/ptx-jni/target/windows-amd64-release/sources/source/PTXStream.cpp:403:101: error: cannot convert 'long int*' to 'jlong*' {aka 'long long int*'}
```

Could you please take a look at this? 
Also, could you comment on the following changes (at least try on Linux and confirm they are working):
1. `posix_memalign`  replaced with `_aligned_malloc`
2.  CMake flag `-rdynamic` replaced with `-export-dynamic`

**Comment by jjfumero** [comment]
Thank you @vsilaev. I will test this also on Linux and on the FPGAs. 

Personally, I can work on it after 31st January. So, you can expect comments/approval from 1st Feb. 

Thanks

**Comment by eix128** [comment]
@vsilaev, thanks for windows support.
Also it will be good if someone adds Android and IOS with RoboVM :)

But for windows these files generated by mingw are not exe file.
They are mingw compatible.
Can we compile native windows exe ? if yes , how ?

**Comment by vsilaev** [comment]
> But for windows these files generated by mingw are not exe file.
> They are mingw compatible.
> Can we compile native windows exe ? if yes , how ?

Well, the build creates native `*.SO` files on Unix/Mac and native `*.DLL` for Windows. There are no EXE files on either platform.
And what do you mean by "mingw compatible"? Mingw is necessary only for build, afterwards you can use TornadoVM as regular Java application with OS-specific native libraries.

**Comment by eix128** [comment]
I mean MingW Command
Administrator@DESKTOP-7HR7L73 MINGW64 /d/Works/TornadoVM
# tornado uk.ac.manchester.tornado.examples.compute.MatrixMultiplication2D 512
Computing MxM of 512x512
        CPU Execution: 0.52 GFlops, Total time = 515 ms
        GPU Execution: 38.35 GFlops, Total Time = 7 ms
        Speedup: 73.57142857142857x
        Verification true

Administrator@DESKTOP-7HR7L73 MINGW64 /d/Works/TornadoVM
#

**Comment by eix128** [comment]
Ahh sorry tornado is just script :)

#!/usr/bin/env bash

#
# This file is part of Tornado: A heterogeneous programming framework:
# https://github.com/beehive-lab/tornadovm
#
# Copyright (c) 2020, APT Group, Department of Computer Science,
# School of Engineering, The University of Manchester. All rights reserved.
# Copyright (c) 2013-2019, APT Group, School of Computer Science,
# The University of Manchester. All rights reserved.
# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
#
# This code is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 only, as
# published by the Free Software Foundation.
#
# This code is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# version 2 for more details (a copy is included in the LICENSE file that
# accompanied this code).
#
# You should have received a copy of the GNU General Public License version
# 2 along with this work; if not, write to the Free Software Foundation,
# Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
#

if [ -z "${TORNADO_SDK}" ]; then
  echo "Please ensure the TORNADO_SDK environment variable is set correctly"
  exit 0
fi

if [ -z "${JAVA_HOME}" ]; then
  echo "Please ensure the JAVA_HOME environment variable is set correctly"
  exit 0
fi

function print_usage() {
  echo "usage:    tornado [options] class [arg0 ... argN]"
  echo ""
  echo "              -h|--help                       Show this message"
  echo "              -d|--debug                      Enable debug output"
  echo ""
  echo "          Debugging options:"
  echo ""
  echo "              --debug                         Print debugging information"
  echo "              --printBytecodes                Print TornadoVM bytecodes"
  echo "              --igv                           Dump GRAAL IR into IGV"
  echo "              --igvLowTier                    Dump OpenCL Low Tier with IGV"
  echo "              --printKernel                   Print autogenerated OpenCL C Kernel"
  echo "              --ea|--enableassertions         Enable Java assertions"
  echo ""
  echo "          TornadoVM settings:"
  echo ""
  echo "              --printFlags                    Print TornadoVM Java flags and exit"
  echo "              --version                       Print TornadoVM version information and exit"
  echo "              --devices                       Print TornadoVM available devices and exit"
  echo ""
  echo "          TornadoVM display all options:"
  echo ""
  echo "              --options                       Print all TornadoVM options and flags, and exit"
  echo ""
  echo "          Maven:"
  echo ""
  echo "              --mavenLocalInstall             Install all maven packages locally"
  echo ""
  exit 0
}

function print_version() {
  cat "${TORNADO_SDK}/etc/tornado.release"
  exit 0
}

function print_all_options() {
  export COLOR_CYAN='\e[0;36m'
  export COLOR_NC='\e[0m' # No Color
  echo "TornadoVM options"
  echo ""
  echo -e "$COLOR_CYAN Debugging Information: $COLOR_NC"
  echo ""
  echo -e "\tPrint all devices available   : --devices "
  echo -e "\tPrint generated kernel        : -Dtornado.print.kernel=True "
  echo -e "\tPrint kernel debug information: -Dtornado.debug=True "
  echo -e "\tEnable full debug information : -Dtornado.fullDebug=True "
  echo -e "\tDump compiled graphs into IGV : -Dgraal.Dump=*:5 -Dgraal.PrintGraph=Network -Dgraal.PrintCFG=false "
  echo ""
  echo -e "$COLOR_CYAN Define on which device the application should run (by default is 0): $COLOR_NC"
  echo ""
  echo -e "\tChange the driver and device to run the application: -D<taskScheduleName>.<taskName>=<driverID>:<deviceID> (e.g., -Ds0.t0.device=1:2)"
  echo -e "\tChange the driver order: -Dtornado.{ptx,opencl}.priority=X (e.g., -Dtornado.opencl.priority=2)"
  echo ""
  echo -e "$COLOR_CYAN Profiler: $COLOR_NC"
  echo ""
  echo -e "\tEnable online profiler        : -Dtornado.profiler=True "
  echo -e "\tSave profiler information     : -Dtornado.profiler=True -Dtornado.profiler.dump.dir=FILENAME.json"
  echo -e "\tEnable profiler and print at the end of the execution: -Dtornado.profiler=True -Dtornado.log.profiler=True "
  echo ""
  echo -e "$COLOR_CYAN Device Heap: $COLOR_NC"
  echo ""
  echo -e "\tResize the device heap (1GB by default): -Dtornado.heap.allocation=512MB"
  echo ""
  echo -e "$COLOR_CYAN FPGA: $COLOR_NC"
  echo ""
  echo -e "\tLoad the FPGA configuration file: -Dtornado.fpga.conf.file=<file>"
  echo -e "\tGet a precompiled binary        : -Dtornado.precompiled.binary=<path> "
  echo ""
  echo -e "$COLOR_CYAN Optimizations: $COLOR_NC"
  echo ""
  echo -e "\tApply full inlining             : -Dtornado.compiler.fullInlining=True  (False by default)"
  echo ""
}

JAVA_CMD=${JAVA_HOME}/bin/java
JAVA_VERSION_OUTPUT=$("$JAVA_CMD" -version 2>&1)
JAVA_VERSION=$(echo "$JAVA_VERSION_OUTPUT" | awk -F[\"\.] -v OFS=. 'NR==1{print $2,$3}')

if [[ $JAVA_VERSION == "9.0" || $JAVA_VERSION == "10.0" ]]; then
  echo "TornadoVM does not support Java 9 and 10"
  exit 0
fi

TORNADO_FLAGS="-Djava.library.path=${TORNADO_SDK}/lib "
if [ "$JAVA_VERSION" = "1.8" ]; then
  TORNADO_FLAGS="${TORNADO_FLAGS} -Djava.ext.dirs=${TORNADO_SDK}/share/java/tornado"
else
  TORNADO_FLAGS="${TORNADO_FLAGS} --module-path .:${TORNADO_SDK}/share/java/tornado"

  # If extra values for module path have been specified by the user, we must append them
  if [[ $@ == *"--module-path"* ]]; then
    USER_MODULE_PATH=$(echo $@ | sed 's/.*--module-path //' | sed 's/ .*//')

    # Remove the user module path entry from the positional parameters
    for arg in "$@"; do
      shift
      case $arg in
        (--module-path) shift ;;
        (*) set -- "$@" "$arg" ;;
      esac
    done

    TORNADO_FLAGS="${TORNADO_FLAGS}:${USER_MODULE_PATH}"
  fi
fi

printflags=0
printdevices=0
enableAssertions=0

POSITIONAL=()
while [[ $# -gt 0 ]]; do
  key="$1"

  case $key in
  --version)
    print_version
    shift
    ;;
  -d | --debug)
    TORNADO_FLAGS="${TORNADO_FLAGS} -Dtornado.debug=True"
    shift # past argument
    ;;
  --igv)
    TORNADO_FLAGS="${TORNADO_FLAGS} -Dgraal.Dump=*:5 -Dgraal.PrintGraph=Network -Dgraal.PrintCFG=false"
    shift # past argument
    ;;
  --igvLowTier)
    TORNADO_FLAGS="${TORNADO_FLAGS} -Dgraal.Dump=*:1 -Dgraal.PrintGraph=Network -Dgraal.PrintCFG=true -Dtornado.debug.lowtier=True "
    shift
    ;;
  --printKernel)
    TORNADO_FLAGS="${TORNADO_FLAGS} -Dtornado.print.kernel=True "
    shift
    ;;
  --printBytecodes)
    TORNADO_FLAGS="${TORNADO_FLAGS} -Dtornado.print.bytecodes=True "
    shift
    ;;
  --mavenLocalInstall)
    tornadoLocalInstallMaven
    exit 0
    ;;
  --printFlags)
    printflags=1
    shift
    ;;
  --devices)
    printdevices=1
    shift
    ;;
  --options)
    print_all_options
    exit 0
    ;;
  -h | --help)
    print_usage
    shift
    ;;
  -ea | --ea | -enableassertions | --enableassertions)
    enableAssertions=1
    shift
    ;;
  *) # unknown option
    POSITIONAL+=("$1") # save it in an array for later
    shift              # past argument
    ;;
  esac
done
set -- "${POSITIONAL[@]}" # restore positional parameters

PROVIDERS=" \
-Dtornado.load.api.implementation=uk.ac.manchester.tornado.runtime.tasks.TornadoTaskSchedule \
-Dtornado.load.runtime.implementation=uk.ac.manchester.tornado.runtime.TornadoCoreRuntime \
-Dtornado.load.tornado.implementation=uk.ac.manchester.tornado.runtime.common.Tornado \
-Dtornado.load.device.implementation.opencl=uk.ac.manchester.tornado.drivers.opencl.runtime.OCLDeviceFactory \
-Dtornado.load.device.implementation.ptx=uk.ac.manchester.tornado.drivers.ptx.runtime.PTXDeviceFactory \
-Dtornado.load.annotation.implementation=uk.ac.manchester.tornado.annotation.ASMClassVisitor \
-Dtornado.load.annotation.parallel=uk.ac.manchester.tornado.api.annotations.Parallel "

JAVA_FLAGS=""
if [ $enableAssertions -eq 1 ]; then
  # We do not satisfy the Graal compiler assertions because we only support a subset of the Java specification.
  # This allows us have the GraalIR in states which normally would be illegal.
  JAVA_FLAGS="-ea -da:org.graalvm.compiler... "
fi

JAVA_FLAGS="${JAVA_FLAGS}-server -XX:-UseCompressedOops -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI ${TORNADO_FLAGS} ${PROVIDERS} "
if [ "$JAVA_VERSION" = "1.8" ]; then
  JAVA_FLAGS="${JAVA_FLAGS} -XX:-UseJVMCIClassLoader "
else
  # We have a Java 11+ version
  COMMON_EXPORTS="${TORNADO_SDK}/etc/exportLists/common-exports"
  OPENCL_EXPORTS="${TORNADO_SDK}/etc/exportLists/opencl-exports"
  PTX_EXPORTS="${TORNADO_SDK}/etc/exportLists/ptx-exports"
  UPGRADE_MODULE_PATH="--upgrade-module-path ${TORNADO_SDK}/share/java/graalJars"

  TORNADO_ADD_MODULES="--add-modules ALL-SYSTEM,tornado.runtime,tornado.annotation,tornado.drivers.common"
  PTX_MODULE="tornado.drivers.ptx"
  OPENCL_MODULE="tornado.drivers.opencl"

  JAVA_FLAGS="${JAVA_FLAGS} -XX:+UseParallelOldGC -XX:-UseBiasedLocking @${COMMON_EXPORTS} "
  # If the current Java build is not a GraalVM build, then we need to patch the module path for the compiler
  if [[ ! $JAVA_VERSION_OUTPUT == *"GraalVM"* ]]; then
    JAVA_FLAGS="${JAVA_FLAGS} ${UPGRADE_MODULE_PATH} "
  fi
  read -ra selected_backends < "${TORNADO_SDK}/etc/tornado.backend"
  if [[ ${selected_backends} == *"ptx-backend"* ]]; then
    JAVA_FLAGS="${JAVA_FLAGS} @${PTX_EXPORTS}"
    TORNADO_ADD_MODULES="${TORNADO_ADD_MODULES},${PTX_MODULE}"
  fi
  if [[ ${selected_backends} == *"opencl-backend"* ]]; then
    JAVA_FLAGS="${JAVA_FLAGS} @${OPENCL_EXPORTS}"
    TORNADO_ADD_MODULES="${TORNADO_ADD_MODULES},${OPENCL_MODULE}"
  fi
  JAVA_FLAGS="${JAVA_FLAGS} ${TORNADO_ADD_MODULES}"
fi

if [ $printflags -eq 1 ]; then
  echo "$JAVA_CMD $JAVA_FLAGS"
  exit 0
fi

if [ $printdevices -eq 1 ]; then
  tornado uk.ac.manchester.tornado.drivers.TornadoDeviceQuery verbose
  exit 0
fi

# shellcheck disable=SC2068
${JAVA_CMD} ${JAVA_FLAGS} $@

**Comment by eix128** [comment]
@vsilaev , is there any easy maven script that i can use this dll on maven directly on Windows ?

**Comment by vsilaev** [comment]
Run
```bash
$ tornado --printFlags
```
And use the text returned as a prototype of the command line on Windows or in your IDE.

**Comment by stratika** [comment]
Dear @vsilaev, thanks for your patch. I can confirm that by following your guidelines I managed to run the unit-tests.

I have a question, as I am dealing with something that does not exist in the other OS that we have run TornadoVM and I am I wondering if my CUDA configuration in Windows is not properly set. I have built TornadoVM in Windows 64 bit (using MSYS2) for the following 3 configurations:
1. Only OpenCL backend
2. Only PTX backend
3. Both backends.

For all these configurations (even the 2nd) the reported devices are the OpenCL devices, despite using the CUDA_PATH for the CUDA Toolkit as instructed. See here:
```
$ tornado --devices

Number of Tornado drivers: 1
Total number of OpenCL devices  : 2
Tornado device=0:0
        Intel(R) OpenCL HD Graphics -- Intel(R) UHD Graphics 630
                Global Memory Size: 12.7 GB
                Local Memory Size: 64.0 KB
                Workgroup Dimensions: 3
                Max WorkGroup Configuration: [256, 256, 256]
                Device OpenCL C version: OpenCL C 2.0

Tornado device=0:1
        NVIDIA CUDA -- GeForce GTX 1050 Ti with Max-Q Design
                Global Memory Size: 4.0 GB
                Local Memory Size: 48.0 KB
                Workgroup Dimensions: 3
                Max WorkGroup Configuration: [1024, 1024, 64]
                Device OpenCL C version: OpenCL C 1.2
```

Additionally, when I try some examples with the `--debug` flag that should report the work-items it reports the OpenCL information even if I have only built the PTX backend. For example:
```
$ tornado-test.py --verbose --debug uk.ac.manchester.tornado.unittests.TestHello#testHello

task info: s0.t0
        platform          : Intel(R) OpenCL HD Graphics
        device            : Intel(R) UHD Graphics 630 CL_DEVICE_TYPE_GPU (available)
        dims              : 1
        global work offset: [0]
        global work size  : [8]
        local  work size  : [8, 1, 1]

tornado[  0,  0,  0]> hello
tornado[  1,  0,  0]> hello
tornado[  2,  0,  0]> hello
tornado[  3,  0,  0]> hello
tornado[  4,  0,  0]> hello
tornado[  5,  0,  0]> hello
tornado[  6,  0,  0]> hello
tornado[  7,  0,  0]> hello
Test: class uk.ac.manchester.tornado.unittests.TestHello#testHello
        Running test: testHello                  ................  [PASS]
Test ran: 1, Failed: 0


Total Time (s): 3.369774580001831
```

For the record, this example is run in half time when executed on the same device (OpenCL Integrated GPU) but in Ubuntu 18.04: (Total Time (s): 1.554168701171875).

Could you please confirm if the execution is done on the PTX device in your setup?

**Comment by vsilaev** [comment]
Re-try with removing folders `bin/bin`, `bin/sdk` and `dist` (all relative to TornadoVM root).

**Comment by stratika** [comment]
> Re-try with removing folders `bin/bin`, `bin/sdk` and `dist` (all relative to TornadoVM root) before each build.

This was the problem. For some reason the `make clean` was not removing the previous built. I think we can proceed with the testing of the PR, to see the behaviour on the FPGAs and Ubuntu/MacOS.

```
$ tornado-test.py --verbose --debug uk.ac.manchester.tornado.unittests.TestHello#testHello

Executing: printHello_8
   Blocks: [8, 1, 1]
    Grids: [1, 1, 1]
tornado[0, 0, 0]> hello
tornado[1, 0, 0]> hello
tornado[2, 0, 0]> hello
tornado[3, 0, 0]> hello
tornado[4, 0, 0]> hello
tornado[5, 0, 0]> hello
tornado[6, 0, 0]> hello
tornado[7, 0, 0]> hello
Test: class uk.ac.manchester.tornado.unittests.TestHello#testHello
        Running test: testHello                  ................  [PASS]
Test ran: 1, Failed: 0

Total Time (s): 3.150331497192383
```

Just a comment: The timings are not reliable because I am running multiple processes on my PC.

**Comment by jjfumero** [comment]
Already merged - https://github.com/beehive-lab/TornadoVM/pull/54
